TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,263 @@
+            for (;;)
+            {
+                // OctetLength of j
+                var jLen = Utils.GetOctetLen(j);
+                // Byte representation of j
+                var EJ = Utils.I2OSP(j, jLen);
+                // Combine EJ with the rest of the string
+                combined = Utils.Combine(combined, EJ);
+                // Pass the bytes to H_1
+                byte[] ER = Utils.MGF1_SHA256(combined, keyLength);
+                // Convert from Bytes to BigInteger
+                BigInteger z_i = Utils.OS2IP(ER);
+                // Check if the output is larger or equal to N OR GCD(z_i, N) != 1
+                if (z_i.CompareTo(Modulus) >= 0 || !(z_i.Gcd(Modulus).Equals(BigInteger.One)))
+                {
+                    j++;
+                    continue;
+                }
+                return z_i;
+            }
+        }
+
+        /// <summary>
+        /// Calculates the value of w as specified in the setup.
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="xValues"> List of x_i values</param>
+        /// <param name="k">Security parameter as specified in the setup.</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        internal static void GetW(RsaPubKey pubKey, byte[] psBytes, BigInteger[] xValues, int k, int keyLength, out BigInteger w)
+        {
+            if (xValues == null)
+                throw new ArgumentNullException(nameof(xValues));
+
+            var BigK = xValues.Length;
+
+            // ASN.1 encoding of the PublicKey
+            var keyBytes = pubKey.ToBytes();
+
+            // Encoding of the x_0
+            var ExLen = Utils.GetByteLength(keyLength);
+            var Ex_0 = Utils.I2OSP(xValues[0], ExLen);
+            byte[] ExComb = Ex_0;
+            // Encoding the rest of the x Values
+            for (int i = 1; i < BigK; i++)
+            {
+                var tmp = Utils.I2OSP(xValues[i], ExLen);
+                ExComb = Utils.Combine(ExComb, tmp);
+                Ex_0 = ExComb;
+            }
+            // Concatenating the rest of s
+            var s = Utils.Combine(keyBytes, Utils.Combine(psBytes, ExComb));
+            // Hash the OctetString
+            var BigW = Utils.SHA256(s);
+            // Truncate to k-bits
+            BigW = Utils.TruncateKbits(BigW, k);
+            // Convert to an Integer and return
+            w = Utils.OS2IP(BigW);
+        }
+
+        /// <summary>
+        /// Calculates the value of K as specified in the equation 7 in the setup
+        /// </summary>
+        /// <param name="k">Security parameter specified in the setup</param>
+        internal static void GetK(int k, out int BigK)
+        {
+            BigK = k + 1;
+            return;
+        }
+
+        /// <summary>
```

Comment:
Hmmm I think that means soundness is not working. I.e. There is something wrong with the protocol, not your implementation. 

Tell me what you get when you compite the value 
(N-phi(N)) 2^k / N
Do you get a small number or a big number?